### PR TITLE
Sergey/si 467

### PR DESCRIPTION
**Pull request by sergey-stratoscale**
1. adding tail functionality to strato-log
2. add remote strato-log reading
    adding flags that allow to configure remote user and password
    allowing to create/update conf file parameters if conf file does not exist
3. set localtime as default setting to strato-log

@Stratoscale/systegration PRPRPRPRPRPRPPRPR

### CODE REVIEW

File: py/strato/common/log/log2text.py [hunk 1]
```diff
@@ -227,20 +227,79 @@ def printLogs(logFiles, formatter):
...
+        fileObj = gzip.GzipFile(filePath, mode)
     else:
-        return open(filePath, mode)
-
+        fileObj = open(filePath, mode)
+    if tail > 0:
+        tailFile(fileObj, tail)
+    return fileObj
+
+def tailFile(fileObj, n):
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: baaaaa......

File: py/strato/common/log/log2text.py [hunk 2]
```diff
@@ -311,15 +370,40 @@ def _getDefaultPaths(self):
...
+    parser.add_argument("-u", "--user", type=str, help='set default remote user to connect to', default=None)
+    parser.add_argument("--restoreLocaltimeOffset", action="store_true", help='restore localtime offset to machine\'s offset')
+    parser.add_argument("-t", "--tail", type=int, help='print n last lines only', default=0)
+
+    args, unknown = parser.parse_known_args()
+    ignoreArgs = []
 
     actionHappened = False
 
+    if args.user != None:
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: Please add those 3 (user, password and node) to a separate function: shouldExecuteRemotly (or something similar)

File: py/strato/common/log/log2text.py [hunk 3]
```diff
@@ -227,20 +227,79 @@ def printLogs(logFiles, formatter):
...
+        else:
+            bufferSize += bufferIncrement
+    fileObj.seek(0, 0)
+    return fileObj
+
+def ping(host):
+    return os.system("ping -c 1 %s > /dev/null 2>&1" % host) == 0
+
+def runRemotely(host, ignoreArgs):
+    try:
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: This is strange flow - you already got the args, so why do you parse the file if not needed? And you already opened it in init. So why to read it again?
Its really hard to follow.
1. Check if should run remotely.

File: py/strato/common/log/log2text.py [hunk 4]
```diff
@@ -227,20 +227,79 @@ def printLogs(logFiles, formatter):
...
+                 'StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null %(hostname)s %(command)s | less -r'
+    hostname = "%(user)s@%(hostname)s" % dict(user=configFile.get("defaultRemoteUser", 'root'),
+                                              hostname=findHostname(host))
+
+    os.system(sshCommand % dict(command=command,
+                                hostname=hostname,
+                                password=configFile.get("defaultRemotePassword", 'password')))
+
+def findHostname(host):
+    possibleSuffixes = ['.service.strato', '.node.strato']
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: you can have those inside the configuration as well

File: py/strato/common/log/log2text.py [hunk 5]
```diff
@@ -227,20 +227,79 @@ def printLogs(logFiles, formatter):
...
+        configFile = {}
+        print "Configuration file was not found"
+
+    args = [arg for arg in sys.argv[1:] if arg not in ignoreArgs]
+    command = "strato-log %s" % ' '.join(args)
+
+    sshCommand = 'sshpass -p %(password)s ssh -o ServerAliveInterval=5 -o ServerAliveCountMax=1 -o ' \
+                 'StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null %(hostname)s %(command)s | less -r'
+    hostname = "%(user)s@%(hostname)s" % dict(user=configFile.get("defaultRemoteUser", 'root'),
+                                              hostname=findHostname(host))
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: Please make all the resolution before you generate the string command - it wuold be much easier to follow.

File: py/strato/common/log/log2text.py [hunk 6]
```diff
@@ -227,20 +227,79 @@ def printLogs(logFiles, formatter):
...
+            while linesInBuffer != n:
+                fileObj.readline()
+                linesInBuffer -= 1
+            return fileObj
+        else:
+            bufferSize += bufferIncrement
+    fileObj.seek(0, 0)
+    return fileObj
+
+def ping(host):
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: resolveRemoteHost.
Ping is problematic since it might be blocked.

Please use the follwoing:
http://stackoverflow.com/questions/11618118/python-check-if-a-hostname-is-resolved

File: py/strato/common/log/log2text.py (hunk 3 above)
Comment (sergey-stratoscale) [inline py/strato/common/log/log2text.py]: this function is not in the parser scope (and in it flow there's no call to the parser locally), and the config file might have changed (user or password updated)

**Comment by sergey-stratoscale** [comment]
updated

File: py/strato/common/log/log2text.py [hunk 7]
```diff
@@ -311,15 +384,29 @@ def _getDefaultPaths(self):
...
+    parser.add_argument("--restoreLocaltimeOffset", action="store_true", help='restore localtime offset to machine\'s offset')
+    parser.add_argument("-t", "--tail", type=int, help='print n last lines only', default=0)
+
+    args, unknown = parser.parse_known_args()
+    ignoreArgs = []
 
     actionHappened = False
 
+    if args.node != None:
+        executeRemotely(args)
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: please call the exit() from here and not internally

File: py/strato/common/log/log2text.py [hunk 8]
```diff
@@ -227,20 +230,76 @@ def printLogs(logFiles, formatter):
...
+
+def findHostname(host, possibleResolveSuffixes):
+    if resolveHostname(host):
+        return host
+    else:
+        for suffix in possibleResolveSuffixes:
+            if resolveHostname(host + suffix):
+                return host + suffix
+    print "No reachable host was found for %s" % host
+    exit(1)
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: why are you calling exit internally?

Comment (sergey-stratoscale) [inline py/strato/common/log/log2text.py]: because it's case specific and if i encounter no resolvable hosts i want to stop the process

Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: sure, but better to have single entry and exit point for you code. It's much easier to read.
Noone should guess that exit is called internally.

File: py/strato/common/log/log2text.py [hunk 9]
```diff
@@ -227,19 +230,78 @@ def printLogs(logFiles, formatter):
...
+
+def findHostname(host, possibleResolveSuffixes):
+    if resolveHostname(host):
+        return host
     else:
-        return open(filePath, mode)
+        for suffix in possibleResolveSuffixes:
+            if resolveHostname(host + suffix):
+                return host + suffix
+    return False
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: return None i guess - its important to keep same type of return values

File: py/strato/common/log/log2text.py [hunk 10]
```diff
@@ -227,19 +230,78 @@ def printLogs(logFiles, formatter):
...
+    host = "%(user)s@%(hostname)s" % dict(user=user,
+                                          hostname=hostname)
+
+    sshCommand = 'sshpass -p %(password)s ssh -X -f -o ServerAliveInterval=5 -o ServerAliveCountMax=1 -o ' \
+                 'StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null %(host)s %(command)s | less -r'
+
+    os.system(sshCommand % dict(command=command,
+                                host=host,
+                                password=password))
+
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: return 0?

File: py/strato/common/log/log2text.py [hunk 11]
```diff
@@ -227,19 +230,79 @@ def printLogs(logFiles, formatter):
...
+        print "Configuration file was not found"
+
+    args = [arg for arg in sys.argv[1:] if arg not in ignoreArgs]
+    command = "strato-log %s" % ' '.join(args)
+
+    user = configFile.get("defaultRemoteUser", 'root')
+    password = configFile.get("defaultRemotePassword", 'password')
+
+    possibleResolveSuffixes = configFile.get("possibleResolveSuffixes", [])
+    hostname=findHostname(host, possibleResolveSuffixes)
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: space

File: py/strato/common/log/log2text.py [hunk 12]
```diff
@@ -227,19 +230,79 @@ def printLogs(logFiles, formatter):
...
+    if hostname != None:
```
Comment (rom-stratoscale) [inline py/strato/common/log/log2text.py]: hostname == None
